### PR TITLE
Track logic contracts in SimpleProject

### LINKED ISSUE #72: Track created logic contracts in lightweight project for reuse
The lightweight project from #70 should keep track of all logic contracts created, so when a new instance is requested, the same logic contract can be reused. This implies tracking not just the name of the contract, but also its bytecode hash to detect changes. We can track the same info as we do now in the CLI.

### PR DESCRIPTION
**Pull request by spalladino**
Reuses the same logic contract for multiple proxies. Supports tracking bytecode changes to optionally redeploy if the contract class changes. And since we were at it, use upgradeAndCall from Proxy to make atomic upgrades.

Fixes #72

### CODE REVIEW

File: packages/lib/src/project/SimpleProject.js [hunk 1]
```diff
@@ -32,12 +33,12 @@ export default class SimpleProject  {
...
 
-  async upgradeProxy(proxyAddress, contractClass, { initMethod, initArgs, initFrom } = {}) {
+  async upgradeProxy(proxyAddress, contractClass, { contractName, initMethod, initArgs, initFrom, redeployIfChanged } = {}) {
     proxyAddress = toAddress(proxyAddress)
-    const implementation = await this._deployImplementation(contractClass);    
+    const implementation = await this._getOrDeployImplementation(contractClass, contractName, redeployIfChanged);    
     log.info(`Upgrading proxy to new logic contract at ${implementation.address}`)
     const proxy = Proxy.at(proxyAddress, this.txParams)
-    await proxy.upgradeTo(implementation)    
+    await proxy.upgradeTo(implementation) // TODO: Use upgradeToAndCall!
```
Comment (facuspagnuolo) [inline packages/lib/src/project/SimpleProject.js]: we should do sth like we are doing in `App#upgradeProxy` right? I mean, if there is no `initMethod` we should call `upgradeTo`, if there is then we should call `upgradeToAndCall`, am I right?

File: packages/lib/src/proxy/Proxy.js [hunk 2]
```diff
@@ -21,9 +21,11 @@ export default class Proxy {
     this.txParams = txParams
   }
 
-  async upgradeTo(address) {
+  async upgradeTo(address, migrateData) {
```
Comment (facuspagnuolo) [inline packages/lib/src/proxy/Proxy.js]: awesome

File: packages/lib/package.json [hunk 3]
```diff
@@ -57,6 +57,7 @@
     "openzeppelin-solidity": "~1.10.0",
     "truffle-config": "^1.0.6",
     "truffle-contract": "^3.0.6",
+    "truffle-flattener": "github:alcuadrado/truffle-flattener#master",
```
[the remark this answers is not in the document]
Comment (jcarpanelli) [reply packages/lib/package.json]: @spalladino, the latest released version of this library includes module exporting. We should be ok to switch to `truffle-flattener@1.2.8` now.

**Comment by facuspagnuolo** [comment]
@spalladino the CI is complaining because of the version requested for `truffle-flattener` I think

**Comment by spalladino** [comment]
Forcing merge as Travis has passed but Github hasn't noticed

**Pull request merged**